File: app/src/main/java/echo/com/surveys/view/CustomViewPager.java
package echo.com.surveys.view;

import android.content.Context;
import android.support.v4.view.MotionEventCompat;
import android.util.AttributeSet;
import android.view.MotionEvent;
import fr.castorflex.android.verticalviewpager.VerticalViewPager;

public class CustomViewPager extends VerticalViewPager {

    float mStartDragY;
    OnSwipeOutListener mOnSwipeOutListener;
    protected boolean isPagingEnabled = true;

    public CustomViewPager(Context context) {
        super(context);
    }

    public CustomViewPager(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public void setOnSwipeOutListener(OnSwipeOutListener listener) {
        mOnSwipeOutListener = listener;
    }

    private void onSwipeOutAtStart() {
        if (mOnSwipeOutListener!=null) {
            mOnSwipeOutListener.onSwipeOutAtStart();
        }
    }

    private void onSwipeOutAtEnd() {
        if (mOnSwipeOutListener!=null) {
            mOnSwipeOutListener.onSwipeOutAtEnd();
        }
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        switch(ev.getAction() & MotionEventCompat.ACTION_MASK){
            case MotionEvent.ACTION_DOWN:
                if(isPagingEnabled) {
                    mStartDragY = ev.getY();
                }
                break;
        }
        return this.isPagingEnabled && super.onInterceptTouchEvent(ev);

    }

    @Override
    public boolean onTouchEvent(MotionEvent ev){

        if(getCurrentItem()==0 || getCurrentItem()==getAdapter().getCount()-1){
            final int action = ev.getAction();
            float y = ev.getY();
            switch(action & MotionEventCompat.ACTION_MASK){
                case MotionEvent.ACTION_MOVE:
                    break;
                case MotionEvent.ACTION_UP:
                    if (getCurrentItem()==0 && y>mStartDragY) {
                        if(y > mStartDragY + 150) {
                            if(isPagingEnabled) {
                                onSwipeOutAtStart();
                            }
                        }
                    }
                    if (getCurrentItem()==getAdapter().getCount()-1 && y<mStartDragY){
                        if(y +150< mStartDragY) {
                            if(isPagingEnabled) {
                                onSwipeOutAtEnd();
                            }
                        }
                    }
                    break;
            }
        }else{
            mStartDragY=0;
        }
        return super.onTouchEvent(ev);

    }

    public void setPagingEnabled(boolean isPagingEnabled) {
        this.isPagingEnabled = isPagingEnabled;
    }

    public interface OnSwipeOutListener {
        void onSwipeOutAtStart();
        void onSwipeOutAtEnd();
    }

}

File: app/src/main/java/echo/com/surveys/view/CustomViewPager.kt
package echo.com.surveys.view

import android.content.Context
import android.support.v4.view.MotionEventCompat
import android.util.AttributeSet
import android.view.MotionEvent
import fr.castorflex.android.verticalviewpager.VerticalViewPager

class CustomViewPager : VerticalViewPager {

    internal var mStartDragY: Float = 0.toFloat()
    internal var mOnSwipeOutListener: OnSwipeOutListener? = null
    var isPagingEnabled = true

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    fun setOnSwipeOutListener(listener: OnSwipeOutListener) {
        mOnSwipeOutListener = listener
    }

    private fun onSwipeOutAtStart() {
        if (mOnSwipeOutListener != null) {
            mOnSwipeOutListener!!.onSwipeOutAtStart()
        }
    }

    private fun onSwipeOutAtEnd() {
        if (mOnSwipeOutListener != null) {
            mOnSwipeOutListener!!.onSwipeOutAtEnd()
        }
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        when (ev.action and MotionEventCompat.ACTION_MASK) {
            MotionEvent.ACTION_DOWN -> if (isPagingEnabled) {
                mStartDragY = ev.y
            }
        }
        return this.isPagingEnabled && super.onInterceptTouchEvent(ev)

    }

    override fun onTouchEvent(ev: MotionEvent): Boolean {

        if (currentItem == 0 || currentItem == adapter.count - 1) {
            val action = ev.action
            val y = ev.y
            when (action and MotionEventCompat.ACTION_MASK) {
                MotionEvent.ACTION_MOVE -> {
                }
                MotionEvent.ACTION_UP -> {
                    if (currentItem == 0 && y > mStartDragY) {
                        if (y > mStartDragY + 150) {
                            if (isPagingEnabled) {
                                onSwipeOutAtStart()
                            }
                        }
                    }
                    if (currentItem == adapter.count - 1 && y < mStartDragY) {
                        if (y + 150 < mStartDragY) {
                            if (isPagingEnabled) {
                                onSwipeOutAtEnd()
                            }
                        }
                    }
                }
            }
        } else {
            mStartDragY = 0f
        }
        return super.onTouchEvent(ev)

    }

//    fun setPagingEnabled(isPagingEnabled: Boolean) {
//        this.isPagingEnabled = isPagingEnabled
//    }

    interface OnSwipeOutListener {
        fun onSwipeOutAtStart()
        fun onSwipeOutAtEnd()
    }

}
